source/Condition.h

```diff
@@ -0,0 +1,194 @@
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef RVALUE_H_
+#define RVALUE_H_
+
+
+#include <cmath>
+#include <limits>
+#include <string>
+#include <type_traits>
+
+
+
+// This class stores either:
+//   1. A condition's value and key
+//   2. A literal value (key is empty)
+//
+// The value (V) is first in the template since (nearly?) all Condition
+// classes will use a std::string key (K). Having a template key type
+// allows storage of scope, wstring, etc. in the future without
+// rewriting this class.
+template < class V, class K = std::string >
+class Condition {
+public:
+	typedef V ValueType;
+	typedef K KeyType;
+
+	constexpr Condition() : value(), key() {}
+	explicit constexpr Condition(const V &value) : value(value), key() {}
+	constexpr Condition(const V &value, const K &key) : value(value), key(key) {}
+	~Condition() {}
+
+	template <class V2, class K2>
+	Condition(const Condition<V2,K2> &other);
+
+	template <class V2, class K2>
+	Condition<V,K> &operator = (const Condition<V2,K2> &other);
+
+	// Update the value from a scope that contains it
+	template<class Getter>
+	const ValueType &UpdateConditions(const Getter &getter);
+
+	// Update the value from a scope that contains it, but use
+	// ValueType() if Validator(weight) is false.
+	template<class Getter, class Validator>
+	const ValueType &UpdateConditions(const Getter &getter, Validator validator);
+
+	// Accessors and mutators
+
+	const ValueType &Value() const { return value; }
+	ValueType &Value() { return value; }
+	const KeyType &Key() const { return key; }
+	KeyType &Key() { return key; }
+
+	// Does this Condition come from the same place as the other one?
+	// If it was a condition, the key must be the same (value doesn't matter)
+	// If it was a literal (no key) then the value must be the same.
+	// If one is literal and the other is conditional, the result is false.
+	bool SameOrigin(const Condition<V,K> &o);
+
+	// Does this originate from a condition?
+	bool HasConditions() const { return !key.empty(); }
+
+	// Does this originate from a literal value (ie. 5.071)?
+	bool IsLiteral() const { return key.empty(); }
+
+	// Floating-point values are false if they're within half the
+	// type's precision of 0 while any other types are passed
+	// through static_cast<bool>
+	explicit operator bool() const;
+
+	// Allow the Condition to be treated as its value in most contexts.
+	operator ValueType() const { return value; }
+
+private:
+	ValueType value;
```

Comment:
what values will this take? double, int?

Reply:
Anything numeric should work.

Reply:
yeah but I meant in this specific context
because the goal is not to make overkill classes that can do more than what we want them to do in all PRs we have in mind; so if its just int and double why bother making it a template is the thing

Reply:
I'm adding static assertions.

Reply:
you misunderstand what I'm saying, I don't mean we should limit what it does I mean do we need templates

Reply:
cant we just always store a double and return an int if needed?

Reply:
No. Some classes require specific types, and a `Condition<T>` that behaves exactly like those types. For example, the WeightedList *needs* weights that act like an unsigned: infinite precision, and no negative numbers. Also, converting frequently will slow down the code.

Reply:
okay fair, plz add that justification to the PR description

Reply:
I added a section titled, "Why does Condition have to be a template class with implicit conversions?"

Reply:
good, you can resolve this